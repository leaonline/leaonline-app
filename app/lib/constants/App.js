export const START_UP_DELAY = 1000 // TODO move to Config.js
