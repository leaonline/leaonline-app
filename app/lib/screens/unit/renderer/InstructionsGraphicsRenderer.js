import React, { useCallback, useEffect, useRef, useState } from 'react'
import { Vibration, View } from 'react-native'
import { InstructionAnimations } from '../instructions/InstructionAnimations'
import { useTts, TTSengine } from '../../../components/Tts'
import { createStyleSheet } from '../../../styles/createStyleSheet'
import { Loading } from '../../../components/Loading'

/**
 *
 * @param props
 * @return {JSX.Element}
 * @component
 */
const InstructionsGraphicsRendererOriginal = (props) => {
  const { Tts } = useTts()
  const [loading, setLoading] = useState(false)
  const { text, subtype, color, page } = props
  const Instruction = InstructionAnimations.get(subtype)

  // XXX: we need to hack this using a timeout in order
  // to force- rerender TTS, otherwise we have a weird
  // glitch that the first tts remains across all pages
  useEffect(() => {
    setLoading(true)

    const timer = setTimeout(() => {
      setLoading(false)
    }, 500)

    return () => clearTimeout(timer)
  }, [text, subtype, page])

  const onInstructionPress = useCallback(() => {
    Vibration.vibrate(100)
    TTSengine.speakImmediately(text)
  }, [text])

  if (loading) {
    return (
      <View style={styles.loadContainer}>
        <Loading color={color} />
      </View>
    )
  }

  if (!Instruction) {
    return (
      <View style={styles.container}>
        <Tts iconColor={color} text={text} style={styles.tts} block />
      </View>
    )
  }

  return (
    <View style={styles.container}>
      <Tts ttsText={text} color={color} dontShowText />
      <View style={styles.instructionWrapper}>
        <Instruction color={color} height={100} width={200} onPress={onInstructionPress} />
      </View>
    </View>
  )
}

export const InstructionsGraphicsRenderer = React.memo(InstructionsGraphicsRendererOriginal)

const styles = createStyleSheet({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start'
  },
  loadContainer: {
    alignItems: 'center',
    justifyContent: 'center'
  },
  tts: {
    margin: 0,
    padding: 0
  },
  instructionWrapper: {
    flex: 1,
    paddingLeft: 20,
    paddingTop: 15,
    paddingBottom: 15,
  }
})
