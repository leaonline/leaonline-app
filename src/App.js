import React, { useEffect, useState } from 'react'
import { Log } from './infrastructure/Log'
import { connectMeteor } from './meteor/connect'
import { loginMeteor } from './meteor/loginMeteor'
import { loggedIn } from './meteor/loggedIn'
import { StyleSheet, View } from 'react-native'
import * as Font from 'expo-font'
import AppLoading from 'expo-app-loading'
import * as Speech from 'expo-speech'
import { TTSengine } from './components/Tts'
import Navigator from './navigation/navigator'
import './i18n'

const log = Log.create('App')

const startApp = async () => {
  log('init App')
  log('fetch fonts')
  await fetchFonts()
  log('connect to meteor')
  try {
    await connectMeteor({ endpoint: 'ws://192.168.12.44:8080/websocket' })
  } catch (connectError) {
    Log.error(connectError)
    // what to do here?
  }

  log('Meteor connected, attempt login')
  let loginSuccessful = false

  try {
    const loginStatus = await loginMeteor()
    log('login result:', JSON.stringify(loginStatus))
    loginSuccessful = !loginStatus.failed && loginStatus._id && loginStatus.username
  } catch (loginError) {
    Log.error(loginError)
    // what to do here?
  }

  // if we are logged-in but the user is not available yet, we
  // set an interval and wait for the user to be available
  if (loginSuccessful && !loggedIn()) {
    throw new Error('Implement interval!')
  }
}

/**
 * @private used to load our custom font
 * @return {Promise<void>}
 */
const fetchFonts = async () => {
  return Font.loadAsync({
    semicolon: require('./assets/fonts/SemikolonPlus-Regular.ttf')
  })
}

/**
 * @private stylesheet
 */
const styles = StyleSheet.create({
  screen: {
    flex: 1
  }
})

// inject expo-speech as our current
// speech-synthesis implementation
TTSengine.setSpeech(Speech)

/**
 * Main Application entry point
 * @category Global
 * @component
 * @returns {JSX.Element}
 */
export default function App () {
  const [initialized, setInitialized] = useState(false)
  const [waitTimeout, setTimeoutComplete] = useState(false)

  // use this effect to make the splash screen remain
  // for a few seconds, once the font has been loaded
  useEffect(() => {
    if (initialized) {
      // TODO Timeout variable should be later set in a separate global environment file
      setTimeout(() => {
        setTimeoutComplete(true)
      }, 1000)
    }
  }, [initialized])

  if (!waitTimeout) {
    return (
      <AppLoading
        startAsync={startApp}
        onFinish={() => setInitialized(true)}
        onError={(error) => console.error(error)}
      />
    )
  }

  return (
    <View style={styles.screen}>
      <Navigator loggedIn={!!loggedIn()} />
    </View>
  )
}
