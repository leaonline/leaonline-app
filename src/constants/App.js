export const START_UP_DELAY = 1000