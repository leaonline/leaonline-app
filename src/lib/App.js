import React, { useCallback } from 'react'
import './i18n'
import { useSplashScreen } from './hooks/useSplashScreen'
import { useConnection } from './hooks/useConnection'
import { ErrorMessage } from './components/ErrorMessage'
import { initContexts } from './startup/initContexts'
import { initAppState } from './startup/initAppState'
import { fetchFonts } from './startup/initFonts'
import { initTTs } from './startup/initTTS'
import { Connecting } from './components/Connecting'
import { ViewContainer } from './components/ViewContainer'
import { MainNavigation } from './navigation/MainNavigation'
import { initExceptionHandling } from './startup/initExceptionHandling'
import { CatchErrors } from './components/CatchErrors'
import { initSound } from './startup/initSound'

const initFunction = [
  initExceptionHandling,
  initContexts,
  initAppState,
  fetchFonts,
  initTTs,
  initSound
]

/**
 * Main Application entry point
 * @category Global
 * @component
 * @returns {JSX.Element}
 */
export const App = function App () {
  const { appIsReady, error, onLayoutRootView } = useSplashScreen(initFunction)
  const { connected } = useConnection()
  const renderConnectionStatus = useCallback(() => {
    if (connected) {
      return null
    }
    return (<Connecting />)
  }, [connected])

  // splashscreen is still active...
  if (!appIsReady) { return null }

  if (error) {
    return (
      <CatchErrors>
        <ViewContainer onLayout={onLayoutRootView}>
          <ErrorMessage error={error.message} onLayout={onLayoutRootView} />
        </ViewContainer>
      </CatchErrors>
    )
  }

  return (
    <CatchErrors>
      <MainNavigation onLayout={onLayoutRootView} />
      {renderConnectionStatus()}
    </CatchErrors>
  )
}
