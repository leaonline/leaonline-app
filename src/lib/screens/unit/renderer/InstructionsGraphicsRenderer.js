import React from 'react'
import { View } from 'react-native'
import { InstructionAnimations } from '../instructions/InstructionAnimations'
import { TTSengine, useTts } from '../../../components/Tts'
import { createStyleSheet } from '../../../styles/createStyleSheet'

/**
 *
 * @param props
 * @return {JSX.Element}
 * @component
 */
export const InstructionsGraphicsRenderer = (props) => {
  const { Tts } = useTts()
  const { hash, text } = props
  const Instruction = InstructionAnimations.get(hash)

  if (!Instruction) {
    return (
      <Tts text={text} style={styles.tts} />
    )
  }

  const readText = () => {
    TTSengine.speakImmediately(text)
  }

  return (
    <View style={styles.container}>
      <Tts text={text} color={props.color} dontShowText />
      <Instruction onPress={readText} width={200} />
    </View>
  )
}

const styles = createStyleSheet({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start'
  },
  tts: {
    margin: 0,
    padding: 0
  }
})
