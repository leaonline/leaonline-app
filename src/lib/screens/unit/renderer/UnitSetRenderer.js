import React, { useRef } from 'react'
import { ScrollView, View } from 'react-native'
import { ContentRenderer } from './ContentRenderer'
import { unitCardStyles } from './unitCardStyles'
import { createStyleSheet } from '../../../styles/createStyleSheet'
import { Layout } from '../../../constants/Layout'

/**
 *
 * @param props {object}
 * @param props.unitSetDoc {object} the unitSet doc
 * @return {JSX.Element}
 * @constructor
 */
export const UnitSetRenderer = props => {
  const scrollViewRef = useRef()
  const { unitSetDoc, dimensionColor } = props

  return (
    <ScrollView ref={scrollViewRef} style={styles.scrollView} keyboardShouldPersistTaps='always'>
      <View style={[unitCardStyles, styles.dropShadow]}>
        <ContentRenderer
          elements={unitSetDoc.story}
          keyPrefix='unitSet'
          color={dimensionColor}
        />
      </View>
    </ScrollView>
  )
}

const styles = createStyleSheet({
  scrollView: {
    flexGrow: 1
  },
  dropShadow: Layout.dropShadow()
})
