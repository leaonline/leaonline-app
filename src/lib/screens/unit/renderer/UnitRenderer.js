import React, { useRef, useEffect, useState } from 'react'
import { KeyboardAvoidingView, ScrollView, Vibration, View } from 'react-native'
import { FadePanel } from '../../../components/FadePanel'
import { mergeStyles } from '../../../styles/mergeStyles'
import { LeaText } from '../../../components/LeaText'
import { Icon } from 'react-native-elements'
import { Colors } from '../../../constants/Colors'
import { createStyleSheet } from '../../../styles/createStyleSheet'
import { InstructionsGraphicsRenderer } from './InstructionsGraphicsRenderer'
import { useTranslation } from 'react-i18next'
import { TTSengine, useTts } from '../../../components/Tts'
import { Layout } from '../../../constants/Layout'
import { useKeyboardVisibilityHandler } from '../../../hooks/useKeyboardVisibilityHandler'
import { Sound } from '../../../env/Sound'
import { unitCardStyles } from './unitCardStyles'
import { ContentRenderer } from './ContentRenderer'
import { useItemSubType } from '../useItemSubType'
import { Log } from '../../../infrastructure/Log'
import { isIOS } from '../../../utils/isIOS'

const PureContentRenderer = React.memo(ContentRenderer)
const debug = Log.create('UnitRenderer', 'debug')
/**
 * Renders the Unit, independent of the surrounding
 * environment.
 *
 * Includes all components, necessary to display elements
 * and interact with them but contains no neither logic
 * for loading data, nor on processing and evaluating inputs.
 *
 * Still, keep in mind, that it's not a PureComponent!
 *
 * @param props {object}
 * @component
 */
export const UnitRenderer = props => {
  const [keyboardStatus, setKeyboardStatus] = useState(undefined)
  const [fadeIn, setFadeIn] = useState(-1)
  const { t } = useTranslation()
  const { Tts } = useTts()
  const scrollViewRef = useRef()
  const lastScrollPos = useRef({ x: 0, y: 0 }).current

  const {
    unitDoc,
    dimensionColor,
    page,
    submitResponse,
    showCorrectResponse,
    scoreResult,
    allTrue,
    taskPageAction
  } = props
  const unitId = unitDoc?._id
  const itemSubtype = useItemSubType({ unitDoc, page })

  // We need to know the Keyboard state in order to show or hide elements.
  // For example: In "editing" mode of a writing item we want to hide the "check" button.
  useKeyboardVisibilityHandler(({ status }) => {
    debug('keyboard visibility changed', status)

    if (status === 'shown') {
      setKeyboardStatus('shown')
    }
    if (status === 'hidden') {
      setKeyboardStatus('hidden')
      // scrollViewRef.current?.scrollToEnd({ animated: true })
      scrollViewRef.current?.scrollTo({
        x: lastScrollPos.x,
        y: lastScrollPos.y,
        animated: true
      })
    }
  })

  const updateLastScrollPos = e => {
    const { contentOffset } = e.nativeEvent
    lastScrollPos.x = contentOffset.x
    lastScrollPos.y = contentOffset.y
  }

  // we want to have our cards appearing in intervals
  useEffect(() => {
    let timer1
    let timer2

    if (page === 0) {
      setFadeIn(0)
      timer1 = setTimeout(() => setFadeIn(1), 500)
      timer2 = setTimeout(() => setFadeIn(2), 1000)
    }

    return () => {
      clearTimeout(timer1)
      clearTimeout(timer2)
    }
  }, [page])

  useEffect(() => {
    if (showCorrectResponse) {
      const handleResponse = async () => {
        if (allTrue) {
          Vibration.vibrate(500)
          scrollViewRef.current?.scrollToEnd({ animated: true })
          TTSengine.stop()
          await Sound.play(RIGHT_ANSWER)
        }
        else {
          Vibration.vibrate(100)
          TTSengine.stop()
          await Sound.play(WRONG_ANSWER)
        }
      }
      handleResponse().catch(Log.error)
    }
  }, [showCorrectResponse, allTrue])

  const renderInstructions = () => {
    const instructions = (unitDoc.pages[page]?.instructions ?? unitDoc.instructions)?.[0]

    if (!instructions) {
      return null
    }

    return (
      <FadePanel style={mergeStyles(unitCardStyles, styles.instructionStyles)} visible={fadeIn >= 1}>
        <View style={styles.pageText}>
          <Icon
            testID='info-icon'
            reverse
            color={Colors.gray}
            size={10}
            name='info'
            type='font-awesome-5'
          />
        </View>
        <InstructionsGraphicsRenderer
          subtype={itemSubtype.subtype}
          text={instructions.value}
          color={dimensionColor}
          page={page}
        />
      </FadePanel>
    )
  }

  const renderAllTrue = () => {
    if (!showCorrectResponse || !allTrue) {
      return null
    }

    return (
      <View style={{ ...unitCardStyles, ...styles.allTrue }}>
        <Tts color={Colors.success} align='center' iconColor={Colors.success} text={t('unitScreen.allTrue')} />
        <Icon
          testID='alltrue-icon'
          reverse
          color={Colors.success}
          size={20}
          name='thumbs-up'
          type='font-awesome-5'
        />
      </View>
    )
  }

  const renderFooter = () => {
    if (keyboardStatus === 'shown') {
      return null
    }
    return (
      <FadePanel style={styles.navigationButtons} visible={fadeIn >= 2}>
        {taskPageAction()}
      </FadePanel>
    )
  }

  return (
    <KeyboardAvoidingView keyboardVerticalOffset={50} behavior={isIOS() ? 'padding' : 'position'}>
      <ScrollView
        ref={scrollViewRef}
        onMomentumScrollEnd={updateLastScrollPos}
        contentContainerStyle={styles.scrollView}
        persistentScrollbar
        keyboardDismissMode='none'
        contentInset={{ bottom: 20 }}
        keyboardShouldPersistTaps='always'
        automaticallyAdjustKeyboardInsets
      >
        {/* 1. PART STIMULI */}
        <FadePanel style={mergeStyles(unitCardStyles, dropShadow)} visible={fadeIn >= 0}>
          <PureContentRenderer
            elements={unitDoc.stimuli}
            keyPrefix={`${unitId}-stimuli`}
            dimensionColor={dimensionColor}
          />
        </FadePanel>

        {/* 2. PART INSTRUCTIONS */}
        {renderInstructions()}

        {/* 3. PART TASK PAGE CONTENT */}
        <FadePanel
          style={{ ...unitCardStyles, borderWidth: 3, borderColor: Colors.gray, paddingTop: 0, paddingBottom: 20 }}
          visible={fadeIn >= 2}
        >
          <LeaText style={styles.pageText}>{page + 1} / {unitDoc.pages.length}</LeaText>

          <PureContentRenderer
            elements={unitDoc.pages[page]?.content}
            keyPrefix={`${unitId}-${page}`}
            scoreResult={showCorrectResponse && scoreResult}
            showCorrectResponse={showCorrectResponse}
            dimensionColor={dimensionColor}
            submitResponse={submitResponse}
          />
        </FadePanel>

        {renderAllTrue()}

        {renderFooter()}
      </ScrollView>
    </KeyboardAvoidingView>
  )
}

const RIGHT_ANSWER = 'rightAnswer'
const WRONG_ANSWER = 'wrongAnswer'

Sound.load(RIGHT_ANSWER, () => require('../../../assets/audio/right_answer.wav'))
Sound.load(WRONG_ANSWER, () => require('../../../assets/audio/wrong_answer.mp3'))

const styles = createStyleSheet({
  instructionStyles: {
    ...Layout.dropShadow(),
    paddingTop: 0,
    justifyContent: 'space-between'
  },
  pageText: {
    alignSelf: 'center',
    backgroundColor: Colors.gray,
    color: Colors.white,
    paddingLeft: 2,
    paddingRight: 2,
    paddingTop: 0,
    paddingBottom: 0,
    marginTop: 0,
    marginBottom: 4,
    fontSize: 16,
    borderWidth: 0.5,
    borderColor: Colors.gray
  },
  allTrue: {
    flexDirection: 'row',
    alignItems: 'stretch',
    alignContent: 'flex-start',
    justifyContent: 'space-between',
    flexShrink: 1,
    borderColor: Colors.success,
    backgroundColor: '#eaffee',
    borderWidth: 4
  },
  scrollView: {
    flexGrow: 1
  },
  navigationButtons: {
    paddingTop: 7,
    marginBottom: 15,
    marginLeft: 4,
    marginRight: 4,
    alignItems: 'stretch'
  }
})

const dropShadow = Layout.dropShadow()
