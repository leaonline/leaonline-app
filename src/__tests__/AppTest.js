import React from 'react'

import renderer from 'react-test-renderer'
import { render } from '@testing-library/react-native'
import App from '../App'

const tree = renderer.create(<App />).toJSON()

jest.mock('react-native/Libraries/LogBox/LogBox');

it('renders correctly', () => {
  expect(tree).toMatchSnapshot()
})

test('render welcome text component correctly', () => {
  const rendered = render(<App />)
  const textComponent = rendered.getByTestId('textField')

  expect(textComponent.props.children).toEqual('Test Meteor Application')
})
