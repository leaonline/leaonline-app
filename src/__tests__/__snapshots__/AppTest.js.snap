// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    testID="textField"
  >
    Test Meteor Application
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="Button"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Press This Button
      </Text>
    </View>
  </View>
</View>
`;
